# ML-10 — Content Action Playbook

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.


## 1. Ranked actions + reason codes

*The queue: what to do first, and why, in words a human trusts.*


In [1]:
# --- Colab setup: clone the repo and run the whole pipeline again (everything here is gitignored) ---
import os

REPO_URL = "https://github.com/muzammil-12345/flyrank-ml-internship.git"
REPO_DIR = "/content/flyrank-ml-internship"

if not os.path.isdir(os.path.join(REPO_DIR, "scripts")):
    !git clone -q {REPO_URL} {REPO_DIR}

%cd {REPO_DIR}/scripts
# this makes the queue, the model, and the charts again from scratch
!python3 run_all.py

%cd {REPO_DIR}/work/notebooks
print("cwd is now:", os.getcwd())


/content/flyrank-ml-internship/scripts



▶ Step 1/5 — Prepare features — clean the data, build the feature vector, define the label


Prepared 30,000 rows from 30,000 raw rows
Wrote /content/flyrank-ml-internship/data/processed/refresh_feature_vector.csv



▶ Step 2/5 — Baseline — a transparent hand-written rule to beat


Wrote baseline queue: /content/flyrank-ml-internship/data/processed/baseline_refresh_queue.csv
Top-50 declining rate (full data, not the evaluated holdout Precision@50): 0.340



▶ Step 3/5 — Train — logistic regression, decision tree, random forest (client-holdout split)


Trained 3 models on 30,000 rows
Split strategy: client_holdout
Best model: random_forest
Wrote predictions: /content/flyrank-ml-internship/data/processed/model_predictions.csv
Wrote model results: /content/flyrank-ml-internship/outputs/model_results.json



▶ Step 4/5 — Evaluate — ranked refresh queue, charts, and the Markdown report


Wrote final refresh queue: /content/flyrank-ml-internship/outputs/refresh_queue.csv
Wrote model report: /content/flyrank-ml-internship/outputs/model_report.md
Wrote charts in: /content/flyrank-ml-internship/outputs/charts



▶ Step 5/5 — Report — a shareable PDF summary


Wrote PDF report: /content/flyrank-ml-internship/outputs/flyrank_refresh_model_results.pdf

Pipeline complete
Rows scored: 30,000
Best model: random_forest
Queue: outputs/refresh_queue.csv
Report: outputs/model_report.md
PDF: /content/flyrank-ml-internship/outputs/flyrank_refresh_model_results.pdf


/content/flyrank-ml-internship/work/notebooks
cwd is now: /content/flyrank-ml-internship/work/notebooks


In [2]:
import sys, os
import pandas as pd

# just looking for the repo folder so this works even if i run it from a weird place
def find_repo_root(start='.'):
   path = os.path.abspath(start)
   for i in range(6):
      if os.path.isdir(os.path.join(path, 'scripts')) and os.path.isdir(os.path.join(path, 'outputs')):
         return path
      path = os.path.dirname(path)
   raise FileNotFoundError("cant find the repo, run the pipeline first")

REPO_ROOT = find_repo_root()
queue = pd.read_csv(os.path.join(REPO_ROOT, 'outputs', 'refresh_queue.csv'))
print("how many pages in the queue:", len(queue))
queue.head(10)


how many pages in the queue: 30000


,final_rank,content_id,client_id,final_refresh_score,best_model_name,best_model_probability,baseline_refresh_score,confidence,suggested_action,final_reason_codes,...,word_count,trend_direction,competition_level,content_type,main_intent,age_tier,freshness_tier,word_count_tier,impression_tier,position_tier
0,1,content_1f080331fa2b,client_3fdba35f04,81.734212,random_forest,0.783472,0.844481,high,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|low...,...,1404.0,down,MEDIUM,keyword article,informational,91-180,91-180,1000-2000,good,page_1
1,2,content_d6570c51c9bd,client_3fdba35f04,81.603243,random_forest,0.849842,0.695884,medium,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1362.0,down,MEDIUM,keyword article,informational,91-180,91-180,1000-2000,moderate,striking
2,3,content_6aa43079fb0c,client_3fdba35f04,81.544618,random_forest,0.789490,0.825477,high,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1457.0,down,LOW,keyword article,informational,91-180,91-180,1000-2000,good,page_1
3,4,content_72e800a9c214,client_3fdba35f04,81.169731,random_forest,0.776297,0.842545,high,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1371.0,down,MEDIUM,keyword article,commercial,91-180,91-180,1000-2000,good,page_1
4,5,content_e04eb9549989,client_3fdba35f04,80.957565,random_forest,0.816010,0.749468,medium,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1408.0,down,LOW,keyword article,informational,91-180,91-180,1000-2000,good,page_1
5,6,content_b69288c5e701,client_3fdba35f04,80.798090,random_forest,0.796332,0.787358,high,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1370.0,down,LOW,keyword article,informational,91-180,91-180,1000-2000,good,page_1
6,7,content_9b6df29f7889,client_3fdba35f04,80.650656,random_forest,0.846499,0.673530,high,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1415.0,down,MEDIUM,keyword article,commercial,91-180,91-180,1000-2000,moderate,page_1
7,8,content_ba6f9dfcbca1,client_3fdba35f04,80.432641,random_forest,0.827496,0.708128,medium,refresh,declining_with_demand|model_decline_risk|visib...,...,1362.0,down,HIGH,keyword article,transactional,91-180,91-180,1000-2000,good,page_3_5
8,9,content_4d76cdb3387b,client_3fdba35f04,80.428403,random_forest,0.844030,0.671993,medium,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1554.0,down,HIGH,keyword article,commercial,91-180,91-180,1000-2000,moderate,top_3
9,10,content_b4f35d640b1c,client_3fdba35f04,80.428243,random_forest,0.845325,0.669168,medium,refresh,declining_with_demand|model_decline_risk|visib...,...,1389.0,down,HIGH,keyword article,commercial,91-180,91-180,1000-2000,good,page_3_5


In [3]:
# how many pages got which action, and the average score for each one
action_counts = queue.groupby('suggested_action').agg(
   pages=('content_id', 'count'),
   avg_score=('final_refresh_score', 'mean'),
   avg_impressions_90d=('impressions_90d', 'mean'),
).round(1).sort_values('avg_score', ascending=False)

print("actions sorted by score, highest first:")
action_counts


actions sorted by score, highest first:


,pages,avg_score,avg_impressions_90d
suggested_action,,,
refresh_and_review_engagement,1993,62.8,13473.8
refresh_and_review_ctr,6654,62.7,7266.9
refresh,8188,57.7,2338.4
expand_and_refresh,82,52.1,1584.3
monitor,13083,36.3,4702.8


In [4]:
# checking which reason codes show up the most, this is basically the WHY behind each action
reason_counts = queue['final_reason_codes'].str.split('|').explode().value_counts()
print("most common reasons:")
reason_counts.head(10)


most common reasons:


final_reason_codes
declining_with_demand          13152
visible_model_opportunity      10650
low_ctr_visible_page            9759
ctr_review_candidate            9759
general_refresh_review          9117
model_decline_risk              8427
page_one_decay_risk             7076
low_engagement_visible_page     6508
engagement_review_candidate     6508
thin_visible_page                 82
Name: count, dtype: int64

**how i see the archetypes (basically the types of pages) and what action they get:**

- page is going down AND people are not clicking on it much (reasons like `low_ctr_visible_page`) -> action becomes `refresh_and_review_ctr`. score around 62.7.
- page is going down AND people leave fast / dont scroll (`low_engagement_visible_page`) -> action becomes `refresh_and_review_engagement`. score around 62.8, this is actually the top scoring group.
- page is going down but no clear reason like ctr or engagement -> action becomes just `refresh`. this is the biggest group, 8,188 pages.
- page is thin (not much content) but still gets visitors (`thin_visible_page`) -> action becomes `expand_and_refresh`. smallest group, only 82 pages, this one actually needs new writing not just an edit.
- everything else with no strong signal -> action becomes `monitor`. lowest score, biggest bucket (13,083 pages), basically wait and watch.


In [5]:
# checking if older pages always score higher (i thought they would, turns out no)
freshness_view = queue.groupby('freshness_tier')['final_refresh_score'].agg(['mean', 'count']).round(1)
print("avg score by how stale the page is:")
freshness_view


avg score by how stale the page is:


,mean,count
freshness_tier,,
0-30,45.5,20480
181+,43.1,174
31-90,60.7,175
91-180,59.3,9171


**the thing i learned here:** i thought the oldest pages would always get the highest score, but thats not true. pages that are stale for 31 to 180 days score the highest (around 59 to 61), pages that were JUST updated (0-30 days) score the lowest (45.5) which makes sense, and pages that have been stale for a really really long time (181+ days) actually go back down a bit (43.1), maybe because they already lost most of their traffic so there is not much left to save. so the rule is not "oldest page first", its more like "catch it while its starting to slip but still has traffic".


## 2. Intended use and limits

*Who uses this, for what — and where it stops being valid.*


In [6]:
# pulling the model numbers again so i have real receipts for the limits below, not just guessing
import json
with open(os.path.join(REPO_ROOT, 'outputs', 'model_results.json')) as f:
   model_results = json.load(f)

best_model_name = model_results['best_model']['name']
best_model_metrics = model_results['models'][best_model_name]

print("split used:", model_results['split_strategy'])
print("best model:", best_model_name)
print("precision@50:", best_model_metrics['precision_at_50'])
print("base rate:", model_results['target_positive_rate'])


split used: client_holdout
best model: random_forest
precision@50: 0.74
base rate: 0.5420666666666667


**who is this for.** an editor or seo person at flyrank who has too many pages to check by hand and needs to know which ones to look at first.

**what its actually good for.** its a sorting tool. it tells you the order to check pages in, it does not tell you what to actually write or fix.

**where it stops working.**

- its only trained on one dataset (30,000 rows, 32 clients), a different client might behave totally different.
- this data is just a snapshot in time, not a real experiment, so the model can say a page LOOKS like its declining, but it cannot say refreshing it WILL fix it. i cant claim that.
- precision@50 is 0.740 on a client holdout split, meaning it was tested on clients it never saw before, but this number can slowly go bad over time as real client stuff changes.
- it never sees the actual title, url, or keywords of a page, so it cant tell if a page is outdated because of something specific to that client, its all just numbers.
- low confidence pages (there are 15,000 of them) should barely be trusted, they basically mean "model is not sure", so a human needs to look closer at those.


## 3. Human review + the no-go list

*What a person must check before acting. What should never be automated.*


In [7]:
# how many pages are high vs medium vs low confidence, tells you how much to trust each row
confidence_counts = queue['confidence'].value_counts().reindex(['high', 'medium', 'low'])
print("confidence split:")
confidence_counts


confidence split:


confidence
high       3602
medium    11398
low       15000
Name: count, dtype: int64

**things a human has to check before doing anything:**

- actually open the page and read it, the model never reads the content, it only looks at numbers.
- check if the topic is even still relevant, sometimes a page is dying because the topic itself is dead, refreshing wont help that.
- look at the `confidence` column first, medium and low confidence rows need way more manual checking, dont just trust the ranking blindly.
- know who the actual client is and what they care about, the model has no idea, it just sees a random client id number.

**stuff that should never be automated, ever:**

- never auto-publish or auto-edit a page just because the score is high, a person reads it first, always.
- never auto-delete or auto-deindex a page based on this queue alone.
- never spend money on new content (the `expand_and_refresh` pages) without a person saying yes first.
- never send this straight to a client without someone checking it, the model can still be wrong on a single page even if it looks good on average.


## 4. Monitoring / retrain triggers

*What would tell you the recommendations went stale?*


In [8]:
# a small snapshot of numbers worth checking again next time this pipeline runs
monitoring_snapshot = {
   'precision_at_50': best_model_metrics['precision_at_50'],
   'base_rate': model_results['target_positive_rate'],
   'high_confidence_pages': int((queue['confidence'] == 'high').sum()),
   'monitor_action_pages': int((queue['suggested_action'] == 'monitor').sum()),
   'avg_final_score': round(queue['final_refresh_score'].mean(), 2),
}
monitoring_snapshot


{'precision_at_50': 0.74,
 'base_rate': 0.5420666666666667,
 'high_confidence_pages': 3602,
 'monitor_action_pages': 13083,
 'avg_final_score': np.float64(49.8)}

**signs that things went stale:**

- **precision@50 goes down** when i test it again on fresh holdout data, means the model isnt ranking as good as before.
- **base rate moves a lot** away from the 0.39 to 0.54 range i saw here, means the mix of declining pages changed and the model doesnt know that yet.
- **high confidence pages drop a lot**, means the model is less sure about everything now, something about the data changed.
- **monitor action count keeps growing** run after run, means fewer pages get a clear action, model is losing its edge.
- **basic stuff shifts a lot**, like avg content age or avg impressions being very different than before, means the content mix itself changed.

**my retrain rule, simple version:** if precision@50 drops noticeably below 0.740, or its been 3+ months since last training, retrain before trusting this queue again.


## 5. Exports for the paper

*Write the queue (and any figures you want to reuse) to work/outputs/ — your paper builds on these files.*


In [9]:
# saving the queue into work/outputs, this csv stays out of git on purpose, the notebook makes it again anytime
work_outputs_dir = os.path.join(REPO_ROOT, 'work', 'outputs')
os.makedirs(work_outputs_dir, exist_ok=True)

queue_export_path = os.path.join(work_outputs_dir, 'action_playbook_queue.csv')
queue.to_csv(queue_export_path, index=False)
print("saved:", queue_export_path)


saved: /content/flyrank-ml-internship/work/outputs/action_playbook_queue.csv


In [10]:
# saving a small json with the important numbers, this one i AM going to commit, its the proof for my paper later
playbook_summary = {
   'best_model': best_model_name,
   'split_strategy': model_results['split_strategy'],
   'precision_at_50': best_model_metrics['precision_at_50'],
   'base_rate': model_results['target_positive_rate'],
   'total_pages': int(len(queue)),
   'action_mix': queue['suggested_action'].value_counts().to_dict(),
   'confidence_mix': queue['confidence'].value_counts().to_dict(),
   'no_go_rules': [
      'never auto publish or auto edit just from the score',
      'never auto delete or auto deindex a page from this queue',
      'never spend money on expand_and_refresh without a human saying yes',
      'never send straight to a client without a human checking first',
   ],
}

summary_path = os.path.join(work_outputs_dir, 'playbook_summary.json')
with open(summary_path, 'w') as f:
   json.dump(playbook_summary, f, indent=2)
print("saved:", summary_path)


saved: /content/flyrank-ml-internship/work/outputs/playbook_summary.json


In [11]:
# grabbing 3 charts that already got made and putting copies in work/figures so my paper can reuse them
import shutil

work_figures_dir = os.path.join(REPO_ROOT, 'work', 'figures')
os.makedirs(work_figures_dir, exist_ok=True)

charts_to_reuse = ['action_mix.svg', 'confidence_mix.svg', 'top_reason_codes.svg']
for chart_name in charts_to_reuse:
   src = os.path.join(REPO_ROOT, 'outputs', 'charts', chart_name)
   dst = os.path.join(work_figures_dir, chart_name)
   shutil.copyfile(src, dst)
   print("copied:", chart_name)


copied: action_mix.svg
copied: confidence_mix.svg
copied: top_reason_codes.svg


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.
